### PR TITLE
Update SSR doc to reflect streaming

### PR DESCRIPTION
**Pull request by chrisdmacrae**
#### What kind of changes does this PR include?

- Minor content fixes (broken links, typos, etc.)
- New or updated content
- Translated content
- Changes to the docs site code
- Something else!

#### Description

- Closes # 
- What does this PR change? Give us a brief description.
- Did you change something visual? A before/after screenshot can be helpful.

This adds a really rough introductory explanation of the streaming feature of the SSR integration.

### CODE REVIEW

**Comment by chrisdmacrae** [comment]
@sarah11918 here's a rough start. It could use some love.

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 1]
```diff
@@ -230,4 +230,64 @@ if (!product) {
...
\ No newline at end of file
+The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).
+
+### Streaming
+
+Astro enables HTTP streaming for all pages that are not prerendered. This means that your page's content will be sent incrementally to the page as it resolves async functions and requests.
+
+You can `await` promises in the frontmatter of your page, or you can move those requests down into Astro components the page renders. The page will then send down the HTML it's capable of sending down without waiting for those async components, and then it will start rendering the components in the order they are found in your page's HTML one-by-one.
```
Comment (Yan-Thomas) [inline src/content/docs/en/guides/server-side-rendering.mdx:239]: Nit: no hyphens in "one by one"
```suggestion
You can `await` promises in the frontmatter of your page, or you can move those requests down into Astro components the page renders. The page will then send down the HTML it's capable of sending down without waiting for those async components, and then it will start rendering the components in the order they are found in your page's HTML one by one.
```

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 2]
```diff
@@ -230,4 +230,64 @@ if (!product) {
...
+---
+<html>
+  <head>
+    <title>Astro example</title>
+  </head>
+  <body>
+    <p>{data.example}</p>
+    <p>{otherData.example}</p>
+  </body>
+</html>
```
Comment (Yan-Thomas) [inline src/content/docs/en/guides/server-side-rendering.mdx:256]: Generally, we tend to be a bit more concrete with our code examples, what do you think of making this one functional? 
```suggestion
---
const personResponse = await fetch('https://randomuser.me/api/');
const personData = await personResponse.json();
const randomPerson = personData.results[0];
const factResponse = await fetch('https://catfact.ninja/fact');
const factData = await factResponse.json();
---
<html>
  <head>
    <title>A name and a fact</title>
  </head>
  <body>
    <p>{randomPerson.name.first}</p>
    <p>{factData.fact}</p>
  </body>
</html>
```

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 3]
```diff
@@ -230,4 +230,64 @@ if (!product) {
...
+```
+
+We can move the fetch call into a component to allow the page to send down the shell of the markup (the html, head, and body) before the fetch request is done loading:
+
+```astro title="src/components/Example.astro"
+---
+const data = await fetch('https://example.com')
+---
+<p>{data.example}<p>
+```
```
Comment (Yan-Thomas) [inline src/content/docs/en/guides/server-side-rendering.mdx:266]:
````suggestion
```astro title="src/components/RandomName.astro"
---
const personResponse = await fetch('https://randomuser.me/api/');
const personData = await personResponse.json();
const randomPerson = personData.results[0];
---
<p>{randomPerson.name.first}</p>
```
````

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 4]
```diff
@@ -230,4 +230,64 @@ if (!product) {
...
+
+```astro title="src/components/OtherExample.astro"
+---
+const otherData = await fetch('https://other-example.com')
+---
+<p>{otherData.example}<p>
+```
```
Comment (Yan-Thomas) [inline src/content/docs/en/guides/server-side-rendering.mdx:273]:
````suggestion
```astro title="src/components/RandomFact.astro"
---
const factResponse = await fetch('https://catfact.ninja/fact');
const factData = await factResponse.json();
---
<p>{factData.fact}</p>
```
````

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 5]
```diff
@@ -230,4 +230,64 @@ if (!product) {
...
+---
+<html>
+  <head>
+    <title>Astro example</title>
+  </head>
+  <body>
+    <Example />
+    <OtherExample />
+  </body>
+</html>
```
Comment (Yan-Thomas) [inline src/content/docs/en/guides/server-side-rendering.mdx:290]:
```suggestion
---
import RandomName from '../components/RandomName.astro'
import RandomFact from '../components/RandomFact.astro'
---
<html>
  <head>
    <title>A name and a fact</title>
  </head>
  <body>
    <RandomName />
    <RandomFact />
  </body>
</html>
```

**Review by Yan-Thomas — COMMENTED** [review]
Good job @chrisdmacrae, thanks for taking the time to add this. I added a few suggestions for you, let me know what you think! 🙌

File: src/content/docs/en/guides/server-side-rendering.mdx (hunk 2 above)
Comment (chrisdmacrae) [reply src/content/docs/en/guides/server-side-rendering.mdx:256]: LOVE IT @Yan-Thomas

File: src/content/docs/en/guides/server-side-rendering.mdx (hunk 3 above)
Comment (chrisdmacrae) [reply src/content/docs/en/guides/server-side-rendering.mdx:266]: 👍🏽

File: src/content/docs/en/guides/server-side-rendering.mdx (hunk 4 above)
Comment (chrisdmacrae) [reply src/content/docs/en/guides/server-side-rendering.mdx:273]: 💯

File: src/content/docs/en/guides/server-side-rendering.mdx (hunk 5 above)
Comment (chrisdmacrae) [reply src/content/docs/en/guides/server-side-rendering.mdx:290]: 🥇

**Comment by sarah11918** [comment]
K, so I just unilaterally applied all of @Yan-Thomas 's changes so we have the new version in readable form. (@chrisdmacrae, you would have been welcome to commit the suggestions yourself!) 

Will now read through this myself!

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 6]
```diff
@@ -230,4 +230,70 @@ if (!product) {
...
\ No newline at end of file
+The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).
+
+### Streaming
+
+Astro enables HTTP streaming for all pages that are not prerendered. This means that your page's content will be sent incrementally to the page as it resolves async functions and requests.
+
+You can `await` promises in the frontmatter of your page, or you can move those requests down into Astro components the page renders. The page will then send down the HTML it's capable of sending down without waiting for those async components, and then it will start rendering the components in the order they are found in your page's HTML one by one.
```
Comment (sarah11918) [inline src/content/docs/en/guides/server-side-rendering.mdx:239]:
```suggestion
To control the order in which your page renders content, you can adjust where you `await` promises. 

When you `await` promises in the frontmatter of your page, your initial page rendering waits for this async data. Moving those requests into separate Astro components means that your page around these components can now render without waiting. Then, it will start rendering the async components in the order they are found in your page's HTML, one by one.
```
K, just a rough edit to start and I might have lost some nuance (and might have got something wrong!), but I think something like this hits the use-case harder up front?

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 7]
```diff
@@ -230,4 +230,70 @@ if (!product) {
...
+
+For example, given a page that fetches some data:
```
Comment (sarah11918) [inline src/content/docs/en/guides/server-side-rendering.mdx:241]:
```suggestion
For example, the following page fetches some data in its frontmatter, which delays page rendering:
```

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 8]
```diff
@@ -230,4 +230,70 @@ if (!product) {
...
+    <title>A name and a fact</title>
+  </head>
+  <body>
+    <RandomName />
+    <RandomFact />
+  </body>
+</html>
+```
+
+Now the page will stream in the HTML around the async components first, and then resolve each component one by one, streaming the resulting HTML.
```
Comment (sarah11918) [inline src/content/docs/en/guides/server-side-rendering.mdx:299]:
``````suggestion
To take advantage of Astro's HTML streaming, move the fetch calls into separate components included in the template. This allows the page to send down the shell of the markup (the `<html>`, `<head>`, and `<body>`) before the fetch request is done loading:

```astro title="src/pages/index.astro" {1, 9}
---
import RandomName from '../components/RandomName.astro'
import RandomFact from '../components/RandomFact.astro'
---
<html>
  <head>
    <title>A name and a fact</title>
  </head>
  <body>
    <RandomName />
    <RandomFact />
  </body>
</html>
```
Now the page will stream in the HTML around the async components first, and then resolve each component (shown below) one by one, streaming the resulting HTML.

```astro title="src/components/RandomName.astro"
---
const personResponse = await fetch('https://randomuser.me/api/');
const personData = await personResponse.json();
const randomPerson = personData.results[0];
---
<p>{randomPerson.name.first}</p>
```

```astro title="src/components/RandomFact.astro"
---
const factResponse = await fetch('https://catfact.ninja/fact');
const factData = await factResponse.json();
---
<p>{factData.fact}</p>
```
``````
I like what you've both done with the example here! (I'm half wondering whether this is now a recipe, but...)

I'm not enen *entirely* convinced we need to show the code for the individual components here, to be honest! (That's why my mind jumped to recipe.) What I did here was reorder to show essentially what we're trying to demonstrate: the comparison of what a page looks like when its fetch requests are in frontmatter (causing delays) and what a page looks like when streaming async components can happen. That *might* be all that's necessary here, to get the point across! Not sure, but take a look at how this page now reads with the subtle reordering and see what you think!

**Review by sarah11918 — COMMENTED** [review]
Yay, @chrisdmacrae, this is so timely as I was able to direct someone asking about streaming in Discord TO THIS PR for help!  :raised_hands: 

I've taken a run through this with some... interesting thoughts? :sweat_smile: See what you think! (And fight me on what I've gotten wrong!)

**Comment by tropperstyle** [comment]
As someone who was looking for more details on HTML streaming, this is very helpful and glad to see the docs getting attention!

I was also surprised to find that HTML streaming only works at the component root-level. See https://github.com/withastro/astro/issues/6561

So using a layout component with `<slot/>` will prevent any HTML from flushing until _all_ descendants of the layout resolve.

This seems important to mention here, or even a warning on the [Layouts](https://docs.astro.build/en/core-concepts/layouts/#overview) page

**Review by Jutanium — CHANGES_REQUESTED** [review]
Blocking this PR as I want to add info about including Promises directly in the template, and Team Platform wants to review this PR

File: src/content/docs/en/guides/server-side-rendering.mdx (hunk 8 above)
Comment (chrisdmacrae) [reply src/content/docs/en/guides/server-side-rendering.mdx:301]: @sarah11918 these edits are amazing!

I'll wait on the platform team's review to make sure we've got this right before committing to it.

**Comment by Jutanium** [comment]
[Release where streaming was introduced](https://github.com/withastro/astro/releases/tag/astro%401.0.0-beta.55)

**Review by Jutanium — APPROVED** [review]

**Comment by ohansemmanuel** [comment]
👍🏼 Great potential addition!! 

🙋🏼 Is it worth mentioning here that Astro does not support out-of-order streaming as in libraries like Qwik, Remix, Solid and Marko?

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 9]
```diff
@@ -230,4 +230,100 @@ if (!product) {
...
\ No newline at end of file
+The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).
+
+### Streaming
+
+Since browsers support HTTP streaming, a document will be broken up into chunks, sent over the network in order, and rendered on the page in that order.
+
+Astro additionally uses streaming to help "unblock" data fetching. Before this feature, Astro would wait until all data finished loading before sending HTML. Now, Astro will render and send components over the network until it finds one that `await`s a Promise. It will wait for the component's Promises to resolve, then render and send the component over the network.
```
Comment (ohansemmanuel) [inline src/content/docs/en/guides/server-side-rendering.mdx:239]: > The wording in this paragraph with "... before this feature ..." and "Now, Astro will...", reads like a paragraph referencing a new release. 

Considering Astro's supported this for a [while](https://github.com/withastro/astro/releases/tag/astro%401.0.0-beta.55) now, is it worth rephrasing to something along the lines of: 

> ..."Astro's supported streaming in its templates since version (x.x.x). (continue paragraph as usual while committing "Before this feature" and "Now, ..") ... Astro additionally uses streaming to help "unblock"  ...

Arguably, we may remove the version details and just keep the paragraph without that info ie., as the docs largely reflect the current state of the project (v2 docs) 

> Astro additionally uses streaming to help "unblock" data fetching. Instead of waiting until all data finishes loading before sending HTML, Astro will render and send components over the network until it finds one that `await`s a Promise. It will wait for the component's Promises to resolve, then render and send the component over the network.

Comment (matthewp) [inline src/content/docs/en/guides/server-side-rendering.mdx:239]: I would probably describe how this works "without" streaming rather than saying "before". I don't think a before/after is very interesting to anyone who wasn't around in the "before" and it becomes increasingly irrelevant as time goes on. But the difference in approaches in the approaches themselves is eternal.

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 10]
```diff
@@ -230,4 +230,100 @@ if (!product) {
 ### Server Endpoints
 
 A server endpoint, also known as an **API route**, is a special function exported from a `.js` or `.ts` file within the `src/pages/` folder. 
-The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).
\ No newline at end of file
+The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).
+
+### Streaming
+
+Since browsers support HTTP streaming, a document will be broken up into chunks, sent over the network in order, and rendered on the page in that order.
```
Comment (matthewp) [inline src/content/docs/en/guides/server-side-rendering.mdx:237]: I would rephrase this from the perspective of what the browser does. Something like:

> Browsers consume HTML incrementally; parsing, rendering into the DOM, and painting. This happens as the full HTML document is consumed. This happens whether or not you intentionally stream your HTML. Network conditions can cause large documents to be downloaded slowly.

File: src/content/docs/en/guides/server-side-rendering.mdx (hunk 9 above)
[the remark this answers is not in the document]
Comment (matthewp) [reply src/content/docs/en/guides/server-side-rendering.mdx:239]: Astro doesn't send components over the network at all. I'm confused by this sentence slightly.

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 11]
```diff
@@ -230,4 +230,104 @@ if (!product) {
...
-The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).
\ No newline at end of file
+The function takes an [endpoint context](/en/reference/api-reference/#endpoint-context) and returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). A powerful feature of SSR, API routes are able to securely execute code on the server. To learn more, see our [Endpoints Guide](/en/core-concepts/endpoints/#server-endpoints-api-routes).
+
+### Streaming
+
+Browsers natively support HTTP streaming, where a document is broken up into chunks, sent over the network in order, and rendered on the page in that order.
+
+During this process, browsers consume HTML incrementally: parsing, rendering into the DOM, and painting. This happens whether or not you intentionally stream your HTML. Network conditions can cause large documents to be downloaded slowly.
```
Comment (sarah11918) [inline src/content/docs/en/guides/server-side-rendering.mdx:239]:
```suggestion
During this process, browsers consume HTML incrementally: parsing, rendering into the DOM, and painting. This happens whether or not you intentionally stream your HTML. Network conditions can cause large documents to be downloaded slowly, and waiting for data fetches can block page rendering.
```

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 12]
```diff
@@ -230,4 +230,104 @@ if (!product) {
...
+
+### Streaming
+
+Browsers natively support HTTP streaming, where a document is broken up into chunks, sent over the network in order, and rendered on the page in that order.
+
+During this process, browsers consume HTML incrementally: parsing, rendering into the DOM, and painting. This happens whether or not you intentionally stream your HTML. Network conditions can cause large documents to be downloaded slowly, and waiting for data fetches can block page rendering.
+
+Astro additionally uses streaming to help "unblock" data fetching. Before this feature, Astro would wait until all data finished loading before sending HTML. Now, Astro will render and send components over the network until it finds one that `await`s a Promise. It will wait for the component's Promises to resolve, then render and send the component over the network.
```
Comment (sarah11918) [inline src/content/docs/en/guides/server-side-rendering.mdx:241]:
```suggestion
```
deleting this in favour of starting with a briefer mention in the paragraph above

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 13]
```diff
@@ -230,4 +230,103 @@ if (!product) {
...
+  <body>
+    <h2>A name</h2>
+    <p>{randomPerson.name.first}</p>
+    <h2>A fact</h2>
+    <p>{factData.fact}</p>
+  </body>
+</html>
+```
+
+Moving the `await` calls into smaller components allows you to take advantage of Astro's streaming. Using the following components, Astro can render some HTML first, such as the title, then send the paragraphs when the data is ready.
```
Comment (sarah11918) [inline src/content/docs/en/guides/server-side-rendering.mdx:267]:
```suggestion
Moving the `await` calls into smaller components allows you to take advantage of Astro's streaming. Using the following components to perform the data fetches, Astro can render some HTML first, such as the title, and then the paragraphs when the data is ready.
```

File: src/content/docs/en/guides/server-side-rendering.mdx [hunk 14]
```diff
@@ -230,4 +230,103 @@ if (!product) {
...
+  <body>
+    <h2>A name</h2>
+    <RandomName />
+    <h2>A fact</h2>
+    <RandomFact />
+  </body>
+</html>
+```
+
+The page will now send each part of the page whenever it is ready. Astro will first send the `head`, `body`, and `h1` tags to the browser. It will then wait for `RandomName` to finish loading its Promises, and then send the resulting HTML over the network. It will render the next heading, and then it will wait for `RandomFact` and send its resulting HTML over the browser.
```
Comment (sarah11918) [inline src/content/docs/en/guides/server-side-rendering.mdx:304]:
```suggestion

The Astro page below using these components can render parts of the page sooner. The `<head`>, `<body>`, and `<h1>` tags are no longer blocked by data fetches. The browser will then wait for `RandomName` to finish loading its Promises, and then send the resulting HTML over the network. It will render the next heading, and then it will wait for `RandomFact` and send its resulting HTML over the browser.

``````astro title="src/pages/index.astro"
---
import RandomName from '../components/RandomName.astro'
import RandomFact from '../components/RandomFact.astro'
---
<html>
  <head>
    <title>A name and a fact</title>
  </head>
  <body>
    <h2>A name</h2>
    <RandomName />
    <h2>A fact</h2>
    <RandomFact />
  </body>
</html>
```
``````
Broke up the code samples by swapping the order of explanatory paragraph and `index.astro` file that uses the above components

**Review by matthewp — APPROVED** [review]
lgtm
